apps/hash-ai-worker-ts/src/activities/infer-entities/propose-entities.ts

```diff
@@ -554,85 +477,182 @@ export const proposeEntities = async (params: {
-              });
-            }
-          } catch (err) {
-            logger.error(
-              `Model provided invalid argument to create_entities function. Argument provided: ${stringify(
-                toolCall.input,
-              )}`,
+          logger.info(
+            `Proposed ${validProposedEntities.length} valid additional entities.`,
+          );
+          logger.info(
+            `Proposed ${invalidProposedEntities.length} invalid additional entities.`,
+          );
+
+          /**
+           * Remove the valid entities from the list of entities in progress.
+           */
+          inferenceState.inProgressEntityIds =
+            inferenceState.inProgressEntityIds.filter(
+              (inProgressEntityId) =>
+                !validProposedEntities.some(
+                  ({ entityId }) => entityId === inProgressEntityId,
+                ),
             );
 
+          /**
+           * The agent may have inferred valid entities that we didn't yet ask it for,
+           * in which case we need to mark them as taken from the queue so we don't
+           * ask for them again.
+           */
+          inferenceState.proposedEntitySummaries =
+            inferenceState.proposedEntitySummaries.map(
+              (proposedEntitySummary) => {
+                if (
+                  !proposedEntitySummary.takenFromQueue &&
+                  validProposedEntities.some(
+                    ({ entityId }) =>
+                      entityId === proposedEntitySummary.entityId,
+                  )
+                ) {
+                  return {
+                    ...proposedEntitySummary,
+                    takenFromQueue: true,
+                  };
+                }
+                return proposedEntitySummary;
+              },
+            );
+
+          inferenceState.proposedEntityCreationsByType = Object.entries(
+            validProposedEntitiesByType,
+          ).reduce((prev, [entityTypeId, proposedEntitiesOfType]) => {
+            return {
+              ...prev,
+              [entityTypeId]: [
+                ...(prev[entityTypeId as VersionedUrl] ?? []),
+                ...proposedEntitiesOfType
+                  .filter(
+                    ({ entityId }) =>
+                      // Don't include invalid entities
+                      !invalidProposedEntities.some(
+                        ({
+                          invalidProposedEntity: { entityId: invalidEntityId },
+                        }) => invalidEntityId === entityId,
+                      ) &&
+                      // Ignore entities we've inferred in a previous iteration, otherwise we'll get duplicates
+                      !inferenceState.proposedEntityCreationsByType[
+                        entityTypeId as VersionedUrl
+                      ]?.some(
+                        (existingEntity) =>
+                          existingEntity.entityId === entityId,
+                      ),
+                  )
```

Comment:
Haven't we already applied this filter above, when creating `validProposedEntitiesByType`?

Reply:
Thanks for catching, have addressed as part of https://github.com/hashintel/hash/pull/4346/commits/69901f0d36896118a445d9e47dcede96ee26f939